ext_authz: properly modify query parameters when keys repeat

turettn opened the pull request:
Commit Message:

Fixes a bug when ext_authz attempts to remove any query parameter in the presence of repeated query parameters.

Additional Description:

In the presence of repeated query parameters removing any query parameter would cause the repeated values to be dropped (e.g. dropping `foo` from `a=b&a=c` resulted in `a=b`)

Risk Level: Low

Testing:  Existing unit tests around ext_authz query param manipulation still.  New unit test added which originally failed with the old code, and works with the new code.

Docs Changes: N/A

Release Notes: Included

Platform Specific Features: N/A

Notes:

I left the original QueryParams type in place, and added a QueryParamsV2 class which I used in the ext_authz module as needed.  The main difference is that QueryParams is a `map<string, string>`, where-as QueryParamsV2 is a wrapper around `map<string, vector<string>>`.

There are notes in the code that QueryParams should eventually be replaced with a class.  Assuming people are happy with this approach, I'm happy to submit follow-on PRs converting all usages of QueryParams to QueryParamsV2, and eventually removing the old type from the code-base entirely.

The old code implemented "query_parameters_to_add" as an overwrite if the key was already present (there are unit tests that explicitly test for this).  There is no way to add repeated parameters.  I left this behavior unmodified, though in the long run I would support something similar to [HeaderAppendAction](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/base.proto#envoy-v3-api-enum-config-core-v3-headervalueoption-headerappendaction) to allow users to specify exactly what they want.

Finally, I have no written C++ professionally in quite some time.  Please let me know what needs improvement, and I'm happy to clean it up as necessary.  I'll be out of the office from 8/5 through 8/12, but will happily pick this up once I'm back.

Fixes #28787

yanavlasov commented on changelogs/current.yaml:
```diff
@@ -5,6 +5,9 @@ behavior_changes:
 # *Changes that may cause incompatibilities for some users, but should not for most*
+- area: ext_authz
+  change: |
+    removing any query parameter in the presence of repeated query parameter keys no longer drops the repeats
```
Format checker wants dot at the end of the sentence. https://dev.azure.com/cncf/envoy/_build/results?buildId=144920&view=logs&j=c7de00aa-b0fb-52ba-707b-da6afd918eda&t=2ec0901d-43fb-521a-88ec-473706b56eab&l=32

```suggestion
    removing any query parameter in the presence of repeated query parameter keys no longer drops the repeats.
```

yanavlasov reviewed (changes requested):
/wait

ggreenway reviewed (commented):
This looks great!

It seems like a bug in most cases to blindly store query parameters in a way that doesn't allow repeated keys. Should we completely replace the old QueryParams code with this?

/wait-any

turettn commented:
> That makes sense; plan SGTM. In that case, can you rename from `v2` to something like `multi` or similar?

Certainly - unless someone has a better name, I'll append a commit changing `QueryParamsV2` to `QueryParamsMulti` this afternoon.

tyxia commented on envoy/http/query_params.h:
```diff
@@ -15,6 +18,22 @@ namespace Utility {
 
+class QueryParamsMulti {
+private:
+  std::map<std::string, std::vector<std::string>> data;
```
I am wondering why the sorted map is used here? Considering unordered map if sorting is not needed.

Also,  please use absl map:
unordered : `absl::flat_hash_map` OR `absl::node_hash_map` if you need pointer stability for key (which not seems to be the case here)
ordered: absl::btree_map

tyxia commented on envoy/http/query_params.h:
```diff
@@ -15,6 +18,22 @@ namespace Utility {
+  std::map<std::string, std::vector<std::string>> data;
+
+public:
+  void remove(std::string key);
```
absl::string_view rather than std::string for `remove`, `overwrite` and `add`? to avoid the copy during the argument passing

turettn replied:
Please bear with me here, because my C++ is rusty at best.

I swapped the `key` parameters to `absl::string_view`s without an issue.  This makes perfect sense - thank you!

I don't think I can safely store `string_view`s in the map, since I have no way to guarantee that the data underlying the `string_view` won't change.

I couldn't come up with a way to avoid copying the `string_view` into a `string` when calling either `add` or `overwrite`, or forming a vector with the string in it.  Typing the parameter as `std::string` seemed performance-equivalent and far easier to read to me.

I tried a bunch of things along the lines of:
```c++
void add(absl::string_view key, absl::string_view value) {
  this->data[key] = std::vector<std::string>{value}; 
}
```
except everything I tried to get `value` to a `std::string` to initialize the vector with made the compiler angry. 

If I'm incorrect, and there is a performance optimization here I'm missing, any pointers you could provide about what the code should look like would be great!

tyxia replied:
Yes, pass by string_view (or const &) and extend its lifetime if needed like storing into container is what I meant.
Thanks!

tyxia reviewed (commented):
Thanks for contribution!

A few quick comments around interface and data structure.

turettn commented:
@tyxia I think that last commit addresses your interface/data structure comments.  Please let me know if any further changes are required.

tyxia commented on envoy/http/query_params.h:
```diff
@@ -15,6 +19,22 @@ namespace Utility {
+
+public:
+  void remove(absl::string_view key);
+  void add(absl::string_view key, const std::string& value);
```
Sorry, I probably was not very clear. What I was thinking is here interface is `void add(absl::string_view key, const absl::string_view value);` and when we need to store the string, we copy it to extend its lifetime (Please see my comment below:)

Same for override() case

tyxia commented on source/common/http/utility.cc:
```diff
@@ -546,6 +567,48 @@ Utility::QueryParams Utility::parseParameters(absl::string_view data, size_t sta
+void Utility::QueryParamsMulti::remove(absl::string_view key) { this->data.erase(key); }
+
+void Utility::QueryParamsMulti::add(absl::string_view key, const std::string& value) {
+  auto result = this->data.emplace(key, std::vector<std::string>{value});
```
For storing in container, it seems safer to extend the lifetime of both key and value.
```
auto result = this->data.emplace(std::string(key), std::vector<std::string>{std::string(value)});
```
string_view to string triggers copy but this copy is needed as we need to store them into container. Your initial approach that passes the arg by value  will trigger copy 2 times (1 is pass by value , the other is here store into map), while the new interface is only one time.

turettn replied:
Well, now I feel dumb.  Suggested changes made.

tyxia reviewed (commented):
In general `string_view` is preferred over `const string&` for function argument type.  `string_view` can allow the caller to pick their own type without copy because it has implicit conversion constructors from both const char* and const std::string&.

I have clarified it a bit in comments below.

turettn commented:
Got it!  I'm afraid I'm out-of-office this week and don't have access to my dev box, but I'll make these changes as soon as I get back.

ggreenway commented:
/wait

turettn commented:
There are two CI checks failing.  One is a timeout on coverage checks that I don't have permissions to re-run.  The other appears to be a connection issue on the integration tests, and I don't think it's related to my changes.

Any advice on how to handle these?

tyxia commented:
/retest

tyxia commented:
/retest mobile

tyxia commented:
Looks like retest command doesn't work.

Add @phlax to help here. Thanks!

turettn commented:
@phlax  whatever you did worked - thank you!

@tyxia I think this is good to go.  Let me know if there's any changes you would like made.

ggreenway reviewed (changes requested):
This looks great overall. Can you add some unit tests for `QueryParamsMulti` to validate its api/functionality?

/wait

turettn commented:
@ggreenway 

> Can you add some unit tests for `QueryParamsMulti` to validate its api/functionality?

For all the existing tests against `parseQueryString`, `parseAndDecodeQueryString`, and `replaceQueryString` in `Envoy::Http::Utility`, I added an equivalent test that called the corresponding method on `QueryParamsMulti`.  I also added tests that exercise these functions with repeating query parameters.  Coverage of the first two provide coverage of `parseParameters`, and coverage of the third provides coverage of `toString`.

These tests are kind of ugly & repetitive right now.  Once I submit all the follow-up PRs swapping `QueryParams` for `QueryParamsMulti`, the tests will shrink back down to their original size (since only the method version of each function will remain).

Finally, I added a new test that exercises the `add`, `remove`, and `overwrite` methods.

tyxia commented on envoy/http/query_params.h:
```diff
@@ -15,6 +19,28 @@ namespace Utility {
 
+class QueryParamsMulti {
+private:
+  absl::btree_map<std::string, std::vector<std::string>> data;
```
coding style nit:  data --- > data_ 

https://github.com/envoyproxy/envoy/blob/main/STYLE.md

tyxia commented on envoy/http/query_params.h:
```diff
@@ -15,6 +19,28 @@ namespace Utility {
+
+  // This function is only here to allow for sane unit testing, and should not be called outside of
+  // unit tests
+  absl::btree_map<std::string, std::vector<std::string>> copy_map_for_testing() {
```
Can we avoid having test specific code in impl? If not, it is probably not a bad idea to just have a getter function. `copy_map_for_testing` is a bit awkward. If this method is needed, we can just call it data(); getData() etc. 
I am not good at naming but a concise function name is preferred IMO

tyxia commented on envoy/http/query_params.h:
```diff
@@ -15,6 +19,28 @@ namespace Utility {
+  // This function is only here to allow for sane unit testing, and should not be called outside of
+  // unit tests
+  absl::btree_map<std::string, std::vector<std::string>> copy_map_for_testing() {
+    return absl::btree_map<std::string, std::vector<std::string>>(data);
```
hmm, we can just `return data;`
Besides, is this function going to be read only, if so, we can 
```
const absl::btree_map<std::string, std::vector<std::string>>&  data() {
    return data;
}
 ```

Generally returning reference raise the concern about lifetime, but here we are returning reference of class member, it is almost safe as people can not access if the class object is destroyed, unless there is a destroy function to clear this member while keeping the class object alive (which is not case here).

tyxia reviewed (commented):
Some coding style nits from me. Thanks!

turettn commented:
Looks like the tests flaked again?

phlax commented:
@turettn could you merge main please - i think the test issue should be resolved

turettn commented:
@ggreenway @tyxia How's this looking?

tyxia reviewed (commented):
Thanks for your contribution! Interface LGTMs. 

I will defer to @ggreenway for review on core functionality (since he already looked at that prat).

ggreenway reviewed (approved):
Looks great! Sorry for the delay in reviewing.
